### PR TITLE
Expose only user-defined types

### PR DESCRIPTION
**Pull request by bsideup**
By default, Guice will register some default types (i.e. Stage, Injector and even `java.util.logging.Logger`). Since we use `getAllBindings()` in GuiceServiceRegistry, we have to make sure that we don't export them. Also, we bind some Furnace-related classes (AddonRegistry, Addon, Furnace, etc), so we don't want to expose them as well.

This change will ensure that we don't export what we shouldn't export.

I also added a test for it to track such unexpected exports.

### CODE REVIEW

File: impl/src/main/java/org/jboss/forge/furnace/container/guice/lifecycle/GuiceAddonLifecycleProvider.java [hunk 1]
```diff
@@ -54,12 +57,14 @@ public void start(Addon addon) throws Exception
...
       modules.add(new GuiceContainerModule(furnace, addonRegistry, addon));
+      modules.add(binder -> binder.bindListener(Matchers.any(), new ServiceTypeListener(addonRegistry)));
+
       for (Module addonModule : ServiceLoader.load(Module.class, addon.getClassLoader()))
       {
          modules.add(addonModule);
       }
 
-      this.injector = Guice.createInjector(modules);
+      this.injector = Guice.createInjector(Stage.PRODUCTION, modules);
```
Comment (bsideup) [inline impl/src/main/java/org/jboss/forge/furnace/container/guice/lifecycle/GuiceAddonLifecycleProvider.java:67]: See https://google.github.io/guice/api-docs/latest/javadoc/index.html?com/google/inject/Stage.html

Since we create an injector for user, we should set production stage for the safety.

File: impl/src/main/java/org/jboss/forge/furnace/container/guice/lifecycle/GuiceAddonLifecycleProvider.java [hunk 2]
```diff
@@ -54,12 +57,14 @@ public void start(Addon addon) throws Exception
    {
       Set<Module> modules = new LinkedHashSet<>();
       modules.add(new GuiceContainerModule(furnace, addonRegistry, addon));
+      modules.add(binder -> binder.bindListener(Matchers.any(), new ServiceTypeListener(addonRegistry)));
```
Comment (bsideup) [inline impl/src/main/java/org/jboss/forge/furnace/container/guice/lifecycle/GuiceAddonLifecycleProvider.java:60]: I hope it's fine to use inline modules for one-liners. Is it fine, according to the codestyle of JBoss?

File: impl/src/main/java/org/jboss/forge/furnace/container/guice/modules/GuiceContainerModule.java [hunk 3]
```diff
@@ -84,6 +77,7 @@ private void bindServices()
                   String line = sc.nextLine();
                   Class serviceType = classLoader.loadClass(line);
                   bind(serviceType).toConstructor(serviceType.getConstructor());
+                  expose(serviceType);
```
Comment (bsideup) [inline impl/src/main/java/org/jboss/forge/furnace/container/guice/modules/GuiceContainerModule.java:80]: PrivateModule will expose only classes explicitly marked as exposed - great feature for our usecase

File: tests/src/test/java/org/jboss/forge/furnace/container/guice/service/AddonRegistryTest.java [hunk 4]
```diff
@@ -57,4 +61,18 @@ public void testServiceInjection() throws Exception
       Assert.assertThat(service.isUnsatisfied(), is(false));
       Assert.assertThat(service.get(), notNullValue());
    }
+
+   @Test
+   public void testExportedTypes() throws Exception
```
Comment (bsideup) [inline tests/src/test/java/org/jboss/forge/furnace/container/guice/service/AddonRegistryTest.java:66]: run this test aside of other changes to see the reason why I did this change. A lot of stuff is exported

File: impl/src/main/java/org/jboss/forge/furnace/container/guice/lifecycle/GuiceAddonLifecycleProvider.java (hunk 2 above)
Comment (gastaldi) [inline impl/src/main/java/org/jboss/forge/furnace/container/guice/lifecycle/GuiceAddonLifecycleProvider.java:60]: Yes, that's fine

**Comment by bsideup** [comment]
@gastaldi I'm wondering if JBoss has some kind of a automated codestyle checker? Since it's non-standard for Java it's a bit hard to follow and I want to make sure I'm not making any mistakes :)

File: tests/src/test/java/org/jboss/forge/furnace/container/guice/service/AddonRegistryTest.java (hunk 4 above)
Comment (gastaldi) [inline tests/src/test/java/org/jboss/forge/furnace/container/guice/service/AddonRegistryTest.java:66]: Cool!

File: impl/src/main/java/org/jboss/forge/furnace/container/guice/modules/GuiceContainerModule.java (hunk 3 above)
Comment (gastaldi) [inline impl/src/main/java/org/jboss/forge/furnace/container/guice/modules/GuiceContainerModule.java:80]: Nice!

File: impl/src/main/java/org/jboss/forge/furnace/container/guice/lifecycle/GuiceAddonLifecycleProvider.java (hunk 1 above)
Comment (gastaldi) [inline impl/src/main/java/org/jboss/forge/furnace/container/guice/lifecycle/GuiceAddonLifecycleProvider.java:67]: Good idea

File: tests/src/test/java/org/jboss/forge/furnace/container/guice/service/AddonRegistryTest.java (hunk 4 above)
Comment (bsideup) [inline tests/src/test/java/org/jboss/forge/furnace/container/guice/service/AddonRegistryTest.java:66]: Maybe we should create some abstract test for the containers to verify that they expose only what is supposed to be exported? Not sure if it's doable because of the semantics of each container, but at least some smoke testing would be good I think

**Comment by gastaldi** [comment]
We thought of using checkstyle some time ago but haven't configured yet. For now, we have the Eclipse Code formatter as a code style requirement: https://github.com/forge/core/blob/master/eclipse-code-formatter-profile.xml

**Comment by bsideup** [comment]
@gastaldi cool! will apply it. I'm using IntelliJ, but I remember some integration with ECF there, shouldn't be a problem :)

File: tests/src/test/java/org/jboss/forge/furnace/container/guice/service/AddonRegistryTest.java (hunk 4 above)
Comment (gastaldi) [inline tests/src/test/java/org/jboss/forge/furnace/container/guice/service/AddonRegistryTest.java:66]: Yeah, that would be a good test to have.
